File: src/Java/BubbleSort_begin.java
import java.util.Arrays;

public class BubbleSort_begin { // Class name matches the filename

    void bubbleSort(int arr[]) {
        int n = arr.length;
        boolean swapped;
        for (int i = 0; i < n - 1; i++) {
            swapped = false;
            for (int j = 0; j < n - i - 1; j++) {
                // TODO: Compare arr[j] and arr[j + 1]
                // If arr[j] is greater than arr[j + 1], they need to be swapped
                // ---- START CODING HERE ----

                // if (arr[j] > arr[j + 1]) {
                //     // swap arr[j+1] and arr[j]
                //     int temp = arr[j];
                //     arr[j] = arr[j + 1];
                //     arr[j + 1] = temp;
                //     swapped = true;
                // }

                // ---- END CODING HERE ----
            }
            // IF no two elements were swapped by inner loop, then break
            if (!swapped)
                break;
        }
    }

    // Method to print the array - already provided
    void printArray(int arr[]) {
        System.out.println(Arrays.toString(arr));
    }

    // Driver method to test above - no changes needed here for the exercise
    public static void main(String args[]) {
        BubbleSort_begin ob = new BubbleSort_begin(); // Use the current class name
        
        System.out.println("Java Begin Version Tests:");
        int arr1[] = {64, 34, 25, 12, 22, 11, 90};
        System.out.println("Java Example 1:");
        System.out.print("Original array: ");
        ob.printArray(arr1);
        ob.bubbleSort(arr1);
        System.out.print("Sorted array: ");
        ob.printArray(arr1);
        System.out.println("--------------------");

        int arr2[] = {5, 1, 4, 2, 8, 3, 7, 6};
        System.out.println("Java Example 2:");
        System.out.print("Original array: ");
        ob.printArray(arr2);
        ob.bubbleSort(arr2);
        System.out.print("Sorted array: ");
        ob.printArray(arr2);
        System.out.println("--------------------");

        int emptyArr[] = {};
        System.out.println("Java Example (Empty Array):");
        System.out.print("Original array: ");
        ob.printArray(emptyArr);
        ob.bubbleSort(emptyArr);
        System.out.print("Sorted array: ");
        ob.printArray(emptyArr);
        System.out.println("--------------------");

        int singleElementArr[] = {42};
        System.out.println("Java Example (Single Element Array):");
        System.out.print("Original array: ");
        ob.printArray(singleElementArr);
        ob.bubbleSort(singleElementArr);
        System.out.print("Sorted array: ");
        ob.printArray(singleElementArr);
    }
}

File: src/C/bubble_sort_begin.c
#include <stdio.h>
#include <stdbool.h>

// Function to swap two elements - already provided
void swap(int *xp, int *yp) {
    int temp = *xp;
    *xp = *yp;
    *yp = temp;
}

// Function to implement bubble sort
void bubbleSort(int arr[], int n) {
    if (n <= 0) return; // Handles empty or invalid size
    int i, j;
    bool swapped;
    for (i = 0; i < n - 1; i++) {
        swapped = false;
        for (j = 0; j < n - i - 1; j++) {
            // TODO: Compare arr[j] and arr[j + 1]
            // If arr[j] is greater than arr[j + 1], they need to be swapped
            // ---- START CODING HERE ----

            // if (arr[j] > arr[j + 1]) {
            //     // Call the swap function
            //     swap(&arr[j], &arr[j + 1]);
            //     swapped = true;
            // }

            // ---- END CODING HERE ----
        }
        // IF no two elements were swapped by inner loop, then break
        if (swapped == false)
            break;
    }
}

// Function to print an array - already provided
void printArray(int arr[], int size) {
    if (size <= 0) {
        printf("\n");
        return;
    }
    int i;
    for (i = 0; i < size; i++)
        printf("%d ", arr[i]);
    printf("\n");
}

// Driver program to test above functions - no changes needed here for the exercise
int main() {
    printf("C Begin Version Tests:\n");
    int arr1[] = {64, 34, 25, 12, 22, 11, 90};
    int n1 = sizeof(arr1) / sizeof(arr1[0]);
    printf("C Example 1:\n");
    printf("Original array: ");
    printArray(arr1, n1);
    bubbleSort(arr1, n1);
    printf("Sorted array: ");
    printArray(arr1, n1);
    printf("--------------------\n");

    int arr2[] = {5, 1, 4, 2, 8, 3, 7, 6};
    int n2 = sizeof(arr2) / sizeof(arr2[0]);
    printf("C Example 2:\n");
    printf("Original array: ");
    printArray(arr2, n2);
    bubbleSort(arr2, n2);
    printf("Sorted array: ");
    printArray(arr2, n2);
    printf("--------------------\n");
    
    int placeholder_for_empty[1]; 
    int n_empty = 0; 
    printf("C Example (Empty Array):\n");
    printf("Original array: ");
    printArray(placeholder_for_empty, n_empty);
    bubbleSort(placeholder_for_empty, n_empty);
    printf("Sorted array: ");
    printArray(placeholder_for_empty, n_empty);
    printf("--------------------\n");

    int single_arr[] = {42};
    int n_single = sizeof(single_arr)/sizeof(single_arr[0]);
    printf("C Example (Single Element Array):\n");
    printf("Original array: ");
    printArray(single_arr, n_single);
    bubbleSort(single_arr, n_single);
    printf("Sorted array: ");
    printArray(single_arr, n_single);

    return 0;
}